### PR TITLE
feat: Iterative deepening search algo

### PR DESCRIPTION
**Pull request by sushskvnitn**
#### Description of Change

This implementation introduces a simplified version of the 8-tile puzzle solver using Iterative Deepening Search. The code has been streamlined to improve clarity and maintainability, while ensuring it remains educational and easy to understand. Unnecessary functions have been removed, and the focus is placed on essential components of the algorithm.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: This implementation introduces a simplified version of the 8-tile puzzle solver using Iterative Deepening Search, enhancing clarity and maintainability while ensuring it remains educational.

### CODE REVIEW

File: search/Iterative_deepening_search_8_tile_puzzle.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
+#include <bits/stdc++.h>
+using namespace std;
```
Comment (realstealthninja) [inline search/Iterative_deepening_search_8_tile_puzzle.cpp:2]:
```suggestion
```
Please do not use bits/stdc++ as it is not portable code
And namespace std is bad practice

Please read the contribution guidelines to see how to contribute effectively

File: search/Iterative_deepening_search_8_tile_puzzle.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+ */
+void takeUserInput(vector<vector<int>> &initialState) {
+    cout << "Enter the initial state of the 8-tile puzzle (0 for blank tile):\n";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            cin >> initialState[i][j];
+        }
+    }
+}
+
```
Comment (realstealthninja) [inline search/Iterative_deepening_search_8_tile_puzzle.cpp:179]: Please add tests

File: search/Iterative_deepening_search_8_tile_puzzle.cpp [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+/**
+ * @brief Main function to run the 8-tile puzzle solver.
+ */
+int main() {
+    vector<vector<int>> initialState(N, vector<int>(N, 0));
+    vector<vector<int>> goalState = {{1, 2, 3}, {4, 5, 6}, {7, 8, 0}};
+
+    takeUserInput(initialState);
+
```
Comment (realstealthninja) [inline search/Iterative_deepening_search_8_tile_puzzle.cpp:188]: Do not have anything else in the main function other than a call to a test function

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Thank you for the contribution

File: search/Iterative_deepening_search_8_tile_puzzle.cpp (hunk 3 above)
Comment (sushskvnitn) [reply search/Iterative_deepening_search_8_tile_puzzle.cpp:182]: did changes in recent PR

File: search/Iterative_deepening_search_8_tile_puzzle.cpp (hunk 1 above)
Comment (sushskvnitn) [reply search/Iterative_deepening_search_8_tile_puzzle.cpp]: did changes in recent PR

File: search/Iterative_deepening_search_8_tile_puzzle.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    // Test Case 1: Solvable case
+    std::vector<std::vector<int>> initialState1 = {{1, 2, 3}, {4, 5, 6}, {7, 8, 0}};
+    std::vector<std::vector<int>> goalState = {{1, 2, 3}, {4, 5, 6}, {7, 8, 0}};
+    
+    Solve8TilePuzzle puzzle1(initialState1, goalState);
+    if (puzzle1.isSolvable()) {
+        std::cout << "Test Case 1: ";
+        puzzle1.iterativeDeepeningSearch();
+    } else {
+        std::cout << "Test Case 1: Not Solvable" << std::endl;
```
Comment (realstealthninja) [inline search/Iterative_deepening_search_8_tile_puzzle.cpp:180]: tests should be checked using assert statements
```cpp
assert(1 == 1) // true
```

File: search/Iterative_deepening_search_8_tile_puzzle.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        for (int j = 0; j < N; j++) {
+            std::cin >> initialState[i][j];
+        }
+    }
+}
+
+/**
+ * @brief Test function to run the 8-tile puzzle solver with predefined cases.
+ */
+void testSolve8TilePuzzle() {
```
Comment (realstealthninja) [inline search/Iterative_deepening_search_8_tile_puzzle.cpp:170]:
```suggestion
static void tests() {
```

File: search/Iterative_deepening_search_8_tile_puzzle.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
+#include <iostream>      // for IO operations
+#include <vector>       // for std::vector
+#include <set>          // for std::set
+#include <limits>       // for std::numeric_limits
+
+/**
+ * @file
```
Comment (realstealthninja) [inline search/Iterative_deepening_search_8_tile_puzzle.cpp:7]: this comment should be before the include statements

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Pull request closed without merging**